File: turtle-soup-bot/src/main/kotlin/io/github/kapu/turtlesoup/service/PuzzleService.kt
package io.github.kapu.turtlesoup.service

import io.github.kapu.turtlesoup.config.PuzzleConfig
import io.github.kapu.turtlesoup.config.PuzzleConstants
import io.github.kapu.turtlesoup.config.PuzzleDedupConstants
import io.github.kapu.turtlesoup.models.Puzzle
import io.github.kapu.turtlesoup.models.PuzzleCategory
import io.github.kapu.turtlesoup.models.PuzzleGenerationRequest
import io.github.kapu.turtlesoup.redis.PuzzleDedupStore
import io.github.kapu.turtlesoup.rest.LlmRestClient
import io.github.kapu.turtlesoup.utils.PuzzleGenerationException
import io.github.kapu.turtlesoup.utils.RedisException
import io.github.oshai.kotlinlogging.KotlinLogging
import io.ktor.client.plugins.ResponseException
import kotlinx.serialization.SerializationException
import java.io.IOException
import java.security.MessageDigest
import java.util.Locale

class PuzzleService(
    private val restClient: LlmRestClient,
    private val puzzleConfig: PuzzleConfig,
    private val dedupStore: PuzzleDedupStore,
) {
    suspend fun generatePuzzle(
        request: PuzzleGenerationRequest,
        chatId: String,
    ): Puzzle {
        val effectiveRequest =
            PuzzleGenerationRequest(
                category = request.category ?: PuzzleCategory.MYSTERY,
                difficulty = request.difficulty ?: PuzzleConstants.DEFAULT_DIFFICULTY,
                theme = request.theme.orEmpty(),
            )
        var lastError: Exception? = null

        repeat(PuzzleDedupConstants.MAX_GENERATION_RETRIES) { attempt ->
            try {
                val puzzle = tryGeneratePuzzle(effectiveRequest, chatId, attempt)
                if (puzzle != null) return puzzle
            } catch (e: ResponseException) {
                lastError = PuzzleGenerationException("Failed to generate puzzle", e)
                logger.warn(e) { "puzzle_generate_failed_response attempt=${attempt + 1}" }
            } catch (e: SerializationException) {
                lastError = PuzzleGenerationException("Failed to generate puzzle", e)
                logger.warn(e) { "puzzle_generate_failed_serialization attempt=${attempt + 1}" }
            } catch (e: IOException) {
                lastError = PuzzleGenerationException("Failed to generate puzzle", e)
                logger.warn(e) { "puzzle_generate_failed_io attempt=${attempt + 1}" }
            } catch (e: RedisException) {
                lastError = PuzzleGenerationException("Failed to generate puzzle", e)
                logger.warn(e) { "puzzle_generate_failed_redis attempt=${attempt + 1}" }
            } catch (e: IllegalStateException) {
                lastError = PuzzleGenerationException("Failed to generate puzzle", e)
                logger.warn(e) { "puzzle_generate_failed_state attempt=${attempt + 1}" }
            }
        }

        val fallback = selectPresetPuzzle(effectiveRequest.difficulty)
        val signature = computeSignature(fallback)
        dedupStore.markUsed(signature, chatId)

        if (lastError != null) {
            logger.info { "puzzle_fallback_preset reason=generate_failed chat_id=$chatId" }
        } else {
            logger.info { "puzzle_fallback_preset reason=duplicate_exhausted chat_id=$chatId" }
        }

        return fallback
    }

    private suspend fun tryGeneratePuzzle(
        request: PuzzleGenerationRequest,
        chatId: String,
        attempt: Int,
    ): Puzzle? {
        val response = restClient.generatePuzzle(request)
        val puzzle = response.toPuzzle()

        if (!puzzle.hasRequiredContent()) {
            logger.warn { "puzzle_invalid_empty_fields attempt=${attempt + 1} chat_id=$chatId" }
            return null
        }

        val signature = computeSignature(puzzle)

        if (dedupStore.isDuplicate(signature, chatId)) {
            logger.warn { "puzzle_duplicate_detected attempt=${attempt + 1} chat_id=$chatId" }
            return null
        }

        dedupStore.markUsed(signature, chatId)
        logger.info { "puzzle_generated chat_id=$chatId difficulty=${puzzle.difficulty}" }
        return puzzle
    }

    suspend fun getRandomPresetPuzzle(): Puzzle {
        val presetResult = restClient.getRandomPuzzle()
        val basePuzzle = presetResult.toPuzzle()
        return applyRewriteIfEnabled(basePuzzle)
    }

    suspend fun getPresetPuzzleByDifficulty(difficulty: Int): Puzzle {
        val presetResult = restClient.getRandomPuzzle(difficulty)
        val basePuzzle = presetResult.toPuzzle()
        return applyRewriteIfEnabled(basePuzzle)
    }

    @Suppress("TooGenericExceptionCaught")
    private suspend fun applyRewriteIfEnabled(puzzle: Puzzle): Puzzle {
        if (!puzzleConfig.rewriteEnabled) {
            logger.info { "preset_puzzle_selected title=${puzzle.title} rewrite=false" }
            return puzzle
        }

        return try {
            logger.info { "rewriting_puzzle title=${puzzle.title}" }
            val rewriteResult =
                restClient.rewriteScenario(
                    title = puzzle.title,
                    scenario = puzzle.scenario,
                    solution = puzzle.solution,
                    difficulty = puzzle.difficulty,
                )
            val rewrittenPuzzle =
                puzzle.copy(
                    scenario = rewriteResult.scenario,
                    solution = rewriteResult.solution,
                )
            logger.info { "puzzle_rewritten title=${puzzle.title}" }
            rewrittenPuzzle
        } catch (
            e: io.ktor.client.plugins.ResponseException,
        ) {
            logger.warn(e) { "rewrite_failed title=${puzzle.title} using_original" }
            puzzle
        } catch (e: kotlinx.serialization.SerializationException) {
            logger.warn(e) { "rewrite_failed_serialization title=${puzzle.title} using_original" }
            puzzle
        } catch (e: java.io.IOException) {
            logger.warn(e) { "rewrite_failed_io title=${puzzle.title} using_original" }
            puzzle
        } catch (e: Exception) {
            logger.warn(e) { "rewrite_failed_unexpected title=${puzzle.title} using_original" }
            puzzle
        }
    }

    private suspend fun selectPresetPuzzle(difficulty: Int?): Puzzle =
        if (difficulty != null) {
            getPresetPuzzleByDifficulty(difficulty)
        } else {
            getRandomPresetPuzzle()
        }

    private fun computeSignature(puzzle: Puzzle): String {
        val normalized =
            listOf(
                puzzle.title,
                puzzle.scenario,
                puzzle.solution,
                puzzle.hints.joinToString("|"),
            ).joinToString("|") { it.trim().lowercase(Locale.ROOT) }

        val digest = MessageDigest.getInstance("SHA-256").digest(normalized.toByteArray())
        return digest.joinToString("") { "%02x".format(it) }
    }

    private fun Puzzle.hasRequiredContent(): Boolean =
        title.isNotBlank() && scenario.isNotBlank() && solution.isNotBlank()

    companion object {
        private val logger = KotlinLogging.logger {}
    }
}

File: game-bot-go/internal/turtlesoup/service/puzzle_service.go
package service

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"log/slog"
	"strings"
	"time"

	"github.com/park285/llm-kakao-bots/game-bot-go/internal/common/llmrest"
	"github.com/park285/llm-kakao-bots/game-bot-go/internal/common/ptr"
	tsconfig "github.com/park285/llm-kakao-bots/game-bot-go/internal/turtlesoup/config"
	tserrors "github.com/park285/llm-kakao-bots/game-bot-go/internal/turtlesoup/errors"
	tsmodel "github.com/park285/llm-kakao-bots/game-bot-go/internal/turtlesoup/model"
	tsredis "github.com/park285/llm-kakao-bots/game-bot-go/internal/turtlesoup/redis"
)

// PuzzleService 는 타입이다.
type PuzzleService struct {
	restClient *llmrest.Client
	cfg        tsconfig.PuzzleConfig
	dedupStore *tsredis.PuzzleDedupStore
	logger     *slog.Logger
}

// NewPuzzleService 는 동작을 수행한다.
func NewPuzzleService(restClient *llmrest.Client, cfg tsconfig.PuzzleConfig, dedupStore *tsredis.PuzzleDedupStore, logger *slog.Logger) *PuzzleService {
	return &PuzzleService{
		restClient: restClient,
		cfg:        cfg,
		dedupStore: dedupStore,
		logger:     logger,
	}
}

// PuzzleGenerationRequest 는 타입이다.
type PuzzleGenerationRequest struct {
	Category   *tsmodel.PuzzleCategory
	Difficulty *int
	Theme      *string
}

// GeneratePuzzle 는 동작을 수행한다.
func (s *PuzzleService) GeneratePuzzle(ctx context.Context, req PuzzleGenerationRequest, chatID string) (tsmodel.Puzzle, error) {
	category := tsmodel.PuzzleCategoryMystery
	if req.Category != nil {
		category = *req.Category
	}

	difficulty := tsconfig.PuzzleDefaultDifficulty
	if req.Difficulty != nil {
		difficulty = *req.Difficulty
	}
	difficulty = clampInt(difficulty, tsconfig.PuzzleMinDifficulty, tsconfig.PuzzleMaxDifficulty)

	theme := ""
	if req.Theme != nil {
		theme = strings.TrimSpace(*req.Theme)
	}

	var lastErr error

	for attempt := 0; attempt < tsconfig.PuzzleDedupMaxGenerationRetries; attempt++ {
		puzzle, err := s.tryGeneratePuzzle(ctx, chatID, category, difficulty, theme, attempt)
		if err != nil {
			lastErr = err
			continue
		}
		return puzzle, nil
	}

	fallback, err := s.getPresetPuzzleByDifficulty(ctx, difficulty)
	if err != nil {
		if lastErr != nil {
			return tsmodel.Puzzle{}, tserrors.PuzzleGenerationError{Err: lastErr}
		}
		return tsmodel.Puzzle{}, err
	}

	signature := computeSignature(fallback)
	_ = s.dedupStore.MarkUsed(ctx, signature, chatID)

	if lastErr != nil {
		s.logger.Info("puzzle_fallback_preset", "reason", "generate_failed", "chat_id", chatID)
	} else {
		s.logger.Info("puzzle_fallback_preset", "reason", "duplicate_exhausted", "chat_id", chatID)
	}

	return fallback, nil
}

func (s *PuzzleService) tryGeneratePuzzle(
	ctx context.Context,
	chatID string,
	category tsmodel.PuzzleCategory,
	difficulty int,
	theme string,
	attempt int,
) (tsmodel.Puzzle, error) {
	req := llmrest.TurtleSoupPuzzleGenerationRequest{
		Category:   ptr.String(string(category)),
		Difficulty: &difficulty,
	}
	if theme != "" {
		req.Theme = &theme
	}

	res, err := s.restClient.TurtleSoupGeneratePuzzle(ctx, req)
	if err != nil {
		s.logger.Warn("puzzle_generate_failed", "attempt", attempt+1, "chat_id", chatID, "err", err)
		return tsmodel.Puzzle{}, tserrors.PuzzleGenerationError{Err: err}
	}

	puzzle := puzzleFromGeneration(res)
	if !hasRequiredContent(puzzle) {
		s.logger.Warn("puzzle_invalid_empty_fields", "attempt", attempt+1, "chat_id", chatID)
		return tsmodel.Puzzle{}, tserrors.PuzzleGenerationError{Err: fmt.Errorf("empty fields")}
	}

	signature := computeSignature(puzzle)
	dup, err := s.dedupStore.IsDuplicate(ctx, signature, chatID)
	if err != nil {
		s.logger.Warn("puzzle_dedup_check_failed", "attempt", attempt+1, "chat_id", chatID, "err", err)
		return tsmodel.Puzzle{}, fmt.Errorf("puzzle dedup check failed: %w", err)
	}
	if dup {
		s.logger.Warn("puzzle_duplicate_detected", "attempt", attempt+1, "chat_id", chatID)
		return tsmodel.Puzzle{}, tserrors.PuzzleGenerationError{Err: fmt.Errorf("duplicate")}
	}

	if err := s.dedupStore.MarkUsed(ctx, signature, chatID); err != nil {
		s.logger.Warn("puzzle_dedup_mark_failed", "attempt", attempt+1, "chat_id", chatID, "err", err)
		return tsmodel.Puzzle{}, fmt.Errorf("puzzle dedup mark failed: %w", err)
	}

	s.logger.Info("puzzle_generated", "chat_id", chatID, "difficulty", puzzle.Difficulty)
	return puzzle, nil
}

func (s *PuzzleService) getPresetPuzzleByDifficulty(ctx context.Context, difficulty int) (tsmodel.Puzzle, error) {
	difficulty = clampInt(difficulty, tsconfig.PuzzleMinDifficulty, tsconfig.PuzzleMaxDifficulty)

	preset, err := s.restClient.TurtleSoupGetRandomPuzzle(ctx, &difficulty)
	if err != nil {
		return tsmodel.Puzzle{}, tserrors.PuzzleGenerationError{Err: err}
	}

	basePuzzle := puzzleFromPreset(preset)
	return s.applyRewriteIfEnabled(ctx, basePuzzle), nil
}

func (s *PuzzleService) applyRewriteIfEnabled(ctx context.Context, puzzle tsmodel.Puzzle) tsmodel.Puzzle {
	if !s.cfg.RewriteEnabled {
		s.logger.Info("preset_puzzle_selected", "title", puzzle.Title, "rewrite", false)
		return puzzle
	}

	s.logger.Info("rewriting_puzzle", "title", puzzle.Title)

	rewritten, err := s.restClient.TurtleSoupRewriteScenario(ctx, puzzle.Title, puzzle.Scenario, puzzle.Solution, puzzle.Difficulty)
	if err != nil {
		s.logger.Warn("rewrite_failed_using_original", "title", puzzle.Title, "err", err)
		return puzzle
	}

	puzzle.Scenario = rewritten.Scenario
	puzzle.Solution = rewritten.Solution
	s.logger.Info("puzzle_rewritten", "title", puzzle.Title)
	return puzzle
}

func puzzleFromGeneration(res *llmrest.TurtleSoupPuzzleGenerationResponse) tsmodel.Puzzle {
	category := tsmodel.ParsePuzzleCategory(res.Category)
	difficulty := clampInt(res.Difficulty, tsconfig.PuzzleMinDifficulty, tsconfig.PuzzleMaxDifficulty)

	return tsmodel.Puzzle{
		Title:      strings.TrimSpace(res.Title),
		Scenario:   strings.TrimSpace(res.Scenario),
		Solution:   strings.TrimSpace(res.Solution),
		Category:   category,
		Difficulty: difficulty,
		Hints:      res.Hints,
		CreatedAt:  timeNow(),
	}
}

func puzzleFromPreset(res *llmrest.TurtleSoupPuzzlePresetResponse) tsmodel.Puzzle {
	title := ""
	if res.Title != nil {
		title = strings.TrimSpace(*res.Title)
	}
	question := ""
	if res.Question != nil {
		question = strings.TrimSpace(*res.Question)
	}
	answer := ""
	if res.Answer != nil {
		answer = strings.TrimSpace(*res.Answer)
	}
	difficulty := tsconfig.PuzzleDefaultDifficulty
	if res.Difficulty != nil {
		difficulty = *res.Difficulty
	}
	difficulty = clampInt(difficulty, tsconfig.PuzzleMinDifficulty, tsconfig.PuzzleMaxDifficulty)

	return tsmodel.Puzzle{
		Title:      title,
		Scenario:   question,
		Solution:   answer,
		Category:   tsmodel.PuzzleCategoryMystery,
		Difficulty: difficulty,
		CreatedAt:  timeNow(),
	}
}

func hasRequiredContent(puzzle tsmodel.Puzzle) bool {
	return strings.TrimSpace(puzzle.Title) != "" && strings.TrimSpace(puzzle.Scenario) != "" && strings.TrimSpace(puzzle.Solution) != ""
}

func computeSignature(puzzle tsmodel.Puzzle) string {
	normalizedParts := []string{
		strings.ToLower(strings.TrimSpace(puzzle.Title)),
		strings.ToLower(strings.TrimSpace(puzzle.Scenario)),
		strings.ToLower(strings.TrimSpace(puzzle.Solution)),
		strings.ToLower(strings.TrimSpace(strings.Join(puzzle.Hints, "|"))),
	}
	normalized := strings.Join(normalizedParts, "|")

	sum := sha256.Sum256([]byte(normalized))
	return hex.EncodeToString(sum[:])
}

func clampInt(v int, minValue int, maxValue int) int {
	if v < minValue {
		return minValue
	}
	if v > maxValue {
		return maxValue
	}
	return v
}

var timeNow = time.Now
